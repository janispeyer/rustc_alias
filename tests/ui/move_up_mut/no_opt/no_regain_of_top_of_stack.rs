// compile-flags: -Zmir-emit-retag

fn access(value: &mut i32) {
    println!("{}", value);
}

fn regain_top_of_stack(x: &mut i32, y: &mut i32) -> i32 {
    access(&mut *x); // lose top of stack here
    *x = 42; // regain top of stack (not implemtned yet, so it's in the no_opt folder)
    *y = 7;
    return *x;
}

fn main() {
    let mut x = 1;
    let mut y = 2;
    regain_top_of_stack(&mut x, &mut y);
}
