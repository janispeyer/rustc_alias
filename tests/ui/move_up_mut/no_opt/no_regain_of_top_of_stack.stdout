# Analysing main
# Analysing regain_top_of_stack
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] <- Retag([fn entry] _1)
[ 1] [_1, _2] <- Retag([fn entry] _2)
[ 2] [_1, _2] <- StorageLive(_4)
[ 3] [_1, _2] <- StorageLive(_5)
[ 4] [_1, _2] <- StorageLive(_6)
[ 5] [_2] <- _6 = &mut (*_1)
[ 6] [_2] <- Retag(_6)
[ 7] [_2] <- _5 = &mut (*_6)
[ 8] [_2] <- Retag([2phase] _5)
[ 9] [_2] <- _4 = access(move _5) -> bb1

bb1
[ 0] [_2] <- StorageDead(_5)
[ 1] [_2] <- StorageDead(_6)
[ 2] [_2] <- StorageDead(_4)
[ 3] [_2] <- (*_1) = const 42_i32
[ 4] [_2] <- (*_2) = const 7_i32
[ 5] [_2] <- _0 = (*_1)
[ 6] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- Retag([fn entry] _2)
[ 2] {} <- StorageLive(_4)
[ 3] {} <- StorageLive(_5)
[ 4] {} <- StorageLive(_6)
[ 5] {} <- _6 = &mut (*_1)
[ 6] {} <- Retag(_6)
[ 7] {} <- _5 = &mut (*_6)
[ 8] {} <- Retag([2phase] _5)
[ 9] {} <- _4 = access(move _5) -> bb1

bb1
[ 0] {} <- StorageDead(_5)
[ 1] {} <- StorageDead(_6)
[ 2] {} <- StorageDead(_4)
[ 3] {} <- (*_1) = const 42_i32
[ 4] {_2: Span(bb1[4])} <- (*_2) = const 7_i32
[ 5] {_2: Span(bb1[4])} <- _0 = (*_1)
[ 6] {_2: Span(bb1[4])} <- return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
StorageLive(_5)
StorageLive(_6)
_6 = &mut (*_1)
Retag(_6)
_5 = &mut (*_6)
Retag([2phase] _5)
_4 = access(move _5) -> bb1

bb1
StorageDead(_5)
StorageDead(_6)
StorageDead(_4)
(*_1) = const 42_i32
(*_2) = const 7_i32
_0 = (*_1)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
StorageLive(_5)
StorageLive(_6)
_6 = &mut (*_1)
Retag(_6)
_5 = &mut (*_6)
Retag([2phase] _5)
_4 = access(move _5) -> bb1

bb1
StorageDead(_5)
StorageDead(_6)
StorageDead(_4)
(*_1) = const 42_i32
(*_2) = const 7_i32
_0 = (*_1)
return

# Analysing access
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1] <- Retag([fn entry] _1)
[ 1] [_1] <- StorageLive(_2)
[ 2] [_1] <- StorageLive(_3)
[ 3] [_1] <- StorageLive(_4)
[ 4] [_1] <- StorageLive(_5)
[ 5] [_1] <- StorageLive(_6)
[ 6] [_1] <- StorageLive(_7)
[ 7] [_1] <- _16 = const _
[ 8] [_1] <- Retag(_16)
[ 9] [_1] <- _7 = &(*_16)
[10] [_1] <- Retag(_7)
[11] [_1] <- _6 = &(*_7)
[12] [_1] <- Retag(_6)
[13] [_1] <- _5 = move _6 as &[&str] (Pointer(Unsize))
[14] [_1] <- Retag(_5)
[15] [_1] <- StorageDead(_6)
[16] [_1] <- StorageLive(_9)
[17] [_1] <- StorageLive(_10)
[18] [_1] <- StorageLive(_11)
[19] [_1] <- StorageLive(_12)
[20] [_1] <- StorageLive(_13)
[21] [_1] <- StorageLive(_14)
[22] [_1] <- StorageLive(_15)
[23] [_1] <- _15 = &_1
[24] [_1] <- Retag(_15)
[25] [_1] <- _14 = &(*_15)
[26] [_1] <- Retag(_14)
[27] [_1] <- _13 = std::fmt::ArgumentV1::<'_>::new_display::<&mut i32>(move _14) -> bb1

bb1
[ 0] [_1] <- Retag(_13)
[ 1] [_1] <- StorageDead(_14)
[ 2] [_1] <- _12 = [move _13]
[ 3] [_1] <- Retag(_12)
[ 4] [_1] <- StorageDead(_13)
[ 5] [_1] <- _11 = &_12
[ 6] [_1] <- Retag(_11)
[ 7] [_1] <- _10 = &(*_11)
[ 8] [_1] <- Retag(_10)
[ 9] [_1] <- _9 = move _10 as &[std::fmt::ArgumentV1<'_>] (Pointer(Unsize))
[10] [_1] <- Retag(_9)
[11] [_1] <- StorageDead(_10)
[12] [_1] <- _4 = std::fmt::Arguments::<'_>::new_v1(move _5, move _9) -> bb2

bb2
[ 0] [_1] <- Retag(_4)
[ 1] [_1] <- StorageDead(_9)
[ 2] [_1] <- StorageDead(_5)
[ 3] [_1] <- _3 = std::io::_print(move _4) -> bb3

bb3
[ 0] [_1] <- StorageDead(_4)
[ 1] [_1] <- StorageDead(_15)
[ 2] [_1] <- StorageDead(_12)
[ 3] [_1] <- StorageDead(_11)
[ 4] [_1] <- StorageDead(_7)
[ 5] [_1] <- StorageDead(_3)
[ 6] [_1] <- _2 = const ()
[ 7] [_1] <- StorageDead(_2)
[ 8] [_1] <- _0 = const ()
[ 9] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- StorageLive(_2)
[ 2] {} <- StorageLive(_3)
[ 3] {} <- StorageLive(_4)
[ 4] {} <- StorageLive(_5)
[ 5] {} <- StorageLive(_6)
[ 6] {} <- StorageLive(_7)
[ 7] {} <- _16 = const _
[ 8] {} <- Retag(_16)
[ 9] {} <- _7 = &(*_16)
[10] {} <- Retag(_7)
[11] {} <- _6 = &(*_7)
[12] {} <- Retag(_6)
[13] {} <- _5 = move _6 as &[&str] (Pointer(Unsize))
[14] {} <- Retag(_5)
[15] {} <- StorageDead(_6)
[16] {} <- StorageLive(_9)
[17] {} <- StorageLive(_10)
[18] {} <- StorageLive(_11)
[19] {} <- StorageLive(_12)
[20] {} <- StorageLive(_13)
[21] {} <- StorageLive(_14)
[22] {} <- StorageLive(_15)
[23] {} <- _15 = &_1
[24] {} <- Retag(_15)
[25] {} <- _14 = &(*_15)
[26] {} <- Retag(_14)
[27] {} <- _13 = std::fmt::ArgumentV1::<'_>::new_display::<&mut i32>(move _14) -> bb1

bb1
[ 0] {} <- Retag(_13)
[ 1] {} <- StorageDead(_14)
[ 2] {} <- _12 = [move _13]
[ 3] {} <- Retag(_12)
[ 4] {} <- StorageDead(_13)
[ 5] {} <- _11 = &_12
[ 6] {} <- Retag(_11)
[ 7] {} <- _10 = &(*_11)
[ 8] {} <- Retag(_10)
[ 9] {} <- _9 = move _10 as &[std::fmt::ArgumentV1<'_>] (Pointer(Unsize))
[10] {} <- Retag(_9)
[11] {} <- StorageDead(_10)
[12] {} <- _4 = std::fmt::Arguments::<'_>::new_v1(move _5, move _9) -> bb2

bb2
[ 0] {} <- Retag(_4)
[ 1] {} <- StorageDead(_9)
[ 2] {} <- StorageDead(_5)
[ 3] {} <- _3 = std::io::_print(move _4) -> bb3

bb3
[ 0] {} <- StorageDead(_4)
[ 1] {} <- StorageDead(_15)
[ 2] {} <- StorageDead(_12)
[ 3] {} <- StorageDead(_11)
[ 4] {} <- StorageDead(_7)
[ 5] {} <- StorageDead(_3)
[ 6] {} <- _2 = const ()
[ 7] {} <- StorageDead(_2)
[ 8] {} <- _0 = const ()
[ 9] {} <- return


# CFG Before
bb0
Retag([fn entry] _1)
StorageLive(_2)
StorageLive(_3)
StorageLive(_4)
StorageLive(_5)
StorageLive(_6)
StorageLive(_7)
_16 = const _
Retag(_16)
_7 = &(*_16)
Retag(_7)
_6 = &(*_7)
Retag(_6)
_5 = move _6 as &[&str] (Pointer(Unsize))
Retag(_5)
StorageDead(_6)
StorageLive(_9)
StorageLive(_10)
StorageLive(_11)
StorageLive(_12)
StorageLive(_13)
StorageLive(_14)
StorageLive(_15)
_15 = &_1
Retag(_15)
_14 = &(*_15)
Retag(_14)
_13 = std::fmt::ArgumentV1::<'_>::new_display::<&mut i32>(move _14) -> bb1

bb1
Retag(_13)
StorageDead(_14)
_12 = [move _13]
Retag(_12)
StorageDead(_13)
_11 = &_12
Retag(_11)
_10 = &(*_11)
Retag(_10)
_9 = move _10 as &[std::fmt::ArgumentV1<'_>] (Pointer(Unsize))
Retag(_9)
StorageDead(_10)
_4 = std::fmt::Arguments::<'_>::new_v1(move _5, move _9) -> bb2

bb2
Retag(_4)
StorageDead(_9)
StorageDead(_5)
_3 = std::io::_print(move _4) -> bb3

bb3
StorageDead(_4)
StorageDead(_15)
StorageDead(_12)
StorageDead(_11)
StorageDead(_7)
StorageDead(_3)
_2 = const ()
StorageDead(_2)
_0 = const ()
return

# CFG After
bb0
Retag([fn entry] _1)
StorageLive(_2)
StorageLive(_3)
StorageLive(_4)
StorageLive(_5)
StorageLive(_6)
StorageLive(_7)
_16 = const _
Retag(_16)
_7 = &(*_16)
Retag(_7)
_6 = &(*_7)
Retag(_6)
_5 = move _6 as &[&str] (Pointer(Unsize))
Retag(_5)
StorageDead(_6)
StorageLive(_9)
StorageLive(_10)
StorageLive(_11)
StorageLive(_12)
StorageLive(_13)
StorageLive(_14)
StorageLive(_15)
_15 = &_1
Retag(_15)
_14 = &(*_15)
Retag(_14)
_13 = std::fmt::ArgumentV1::<'_>::new_display::<&mut i32>(move _14) -> bb1

bb1
Retag(_13)
StorageDead(_14)
_12 = [move _13]
Retag(_12)
StorageDead(_13)
_11 = &_12
Retag(_11)
_10 = &(*_11)
Retag(_10)
_9 = move _10 as &[std::fmt::ArgumentV1<'_>] (Pointer(Unsize))
Retag(_9)
StorageDead(_10)
_4 = std::fmt::Arguments::<'_>::new_v1(move _5, move _9) -> bb2

bb2
Retag(_4)
StorageDead(_9)
StorageDead(_5)
_3 = std::io::_print(move _4) -> bb3

bb3
StorageDead(_4)
StorageDead(_15)
StorageDead(_12)
StorageDead(_11)
StorageDead(_7)
StorageDead(_3)
_2 = const ()
StorageDead(_2)
_0 = const ()
return

